lib/src/main/java/edu/stanford/futuredata/macrobase/datamodel/DataFrame.java

```diff
@@ -169,35 +169,35 @@ public static DataFrame unionAll(List<DataFrame> others) {
             n += others.get(i).numRows;
         }
         combined.numRows = n;
+        int d = first.schema.getNumColumns();
 
-        int d = first.stringCols.size();
-        combined.stringCols = new ArrayList<>(d);
         for (int colIdx = 0; colIdx < d; colIdx++) {
-            String[] newCol = new String[n];
-            int i = 0;
-            for (DataFrame curOther : others) {
-                String[] otherCol = curOther.getStringColumn(colIdx);
```

Comment:
the bug here was that you need to index columns by their global index, not by the internal index when using these apis.